### ISSUE TITLE
Separate library for keybinding description?

### ISSUE THREAD

**Issue by noctuid** [emacs-evil/evil-collection#157]
@justbur Given the huge similarity between this package's and general's describe keybinding functionality (e.g. both use org and tables), do you think that it would make sense to consolidate the two in a separate library used by both evil-collection and general? I'm not sure how familiar you are (or if you are familiar) with general's describe keybindings functionality, but I could list the differences in formatting/functionality if you want, and then if you think it would make sense to create a separate library, we could go from there. I think this would result in a better and consistent experience from both packages (and potentially others in the future).

**Comment by noctuid** [comment]
> Although it would be nice to distinguish between who made the binding.

Yes, I think this is necessary. Packages should provide a symbol (or string or whatever) when registering keybindings and for the actual describe function. Evil-collection's describe function would then just call the library's describe function with this symbol. The internal formatting could be changed to have this symbol as a key to the hash table of keymaps to keybindings.

General currently uses an alist for storing keybindings, but I don't have a huge preference for the internal formatting. I chose an alist since it can be sorted, but iterating through sorted keys works fine, hash tables scale better, and `gethash` is probably more convenient. It looks like the final level/keybindings list is technically an alist of state to keybindings in this package though, which is basically the same as in general.el.

General additionally shows the previous keybinding for the key (nil if previously unbound). It has two methods of updating the previous keybinding. Either it will only update the definition if it has changed from the previous one (so that multiple calls to a key definer won't clobber the old definition; the default) or only if the key was previously unbound.

General truncates definitions instead of ignoring lambdas and menu-items. I think it would make sense to support either option. It's unfortunate org tables don't really support any form of wrapping. Maybe it might make sense to use a link/footnote to show the full lambda or menu item below the table in an org source block?

General doesn't show the evil state as an entry in the table since some users don't use evil at all. Instead states are shown as subheadings. If there is no state, the table is directly after the top-level heading. I'd be fine with allowing for either functionality as long as there was an option to show states as subheadings.

It also allows the user to specify functions for sorting:
- keymaps (can be used to sort everything below too; I don't think this is possible with nested hash tables, but it's not necessary)
- states
- the actual keybindings

Additionally, it allows for "prioritized keybindings" that are shown above all others. Currently, this is used to show the global evil keymaps and `global` and `local` keybindings at the top. The other keybindings show up in the order they were made, but functions are provided for alphabetic sorting. General does not currently have a way to show only active keybindings.

Do you have any issues with or anything to add about this functionality? I'd be fine using your code as the base since it already works with evil-collection. I can make PRs for my suggested functionality from general if that's acceptable.

As a side note, I've been planning on allowing similar functionality for settings and hooks with general. The library could potentially be made to support generic sorting/display of information in org tables at some point.

> I think you could also think about integrating general into this repo.

We've talked about this before, and I would be fine potentially creating a separate library or core that is intended for use by packages, but I'd need to know what is considered useful. Most of general's functionality makes sense primarily for use in a user's configuration. Macros for generating menu-items seem to be the main thing useful for evil-collection (and those are short). Evil is also only an optional dependency of general, so I'd rather any such library be independent from evil-collection.

> I wonder if general could be of use

I think especially with @justbur's improvements to `evil-define-key(*)` (i.e. its support for `local`, `global`, and mode names to use `evil-define-minor-mode-key`), general doesn't add much that would be useful for evil-collection.

**Comment by justbur** [comment]
> Do you have any issues with or anything to add about this functionality? I'd be fine using your code as the base since it already works with evil-collection. I can make PRs for my suggested functionality from general if that's acceptable.

No issues. That all sounds good to me.

> As a side note, I've been planning on allowing similar functionality for settings and hooks with general. The library could potentially be made to support generic sorting/display of information in org tables at some point.

That's a cool idea. 

> We've talked about this before, and I would be fine potentially creating a separate library or core that is intended for use by packages, but I'd need to know what is considered useful. Most of general's functionality makes sense primarily for use in a user's configuration. Macros for generating menu-items seem to be the main thing useful for evil-collection (and those are short). Evil is also only an optional dependency of general, so I'd rather any such library be independent from evil-collection.

I just meant making general a dependency of evil-collection, and using the general key definers (with the added ability to specify who's making the the binding). I understand that general is user focused and much of this wouldn't be used by evil-collection, but I think it might be helpful to point people towards general as a useful package to use with evil. 

In other words, I don't think you would need to make a different (core) version of general.

**Comment by noctuid** [comment]
Just as a heads up, I've started working on a library for this. I've implemented the heading/table functionality in a generic way that could be used for settings, advice, hooks, or anything else. The library takes care of storage and most of the defaults for display but allows for customizing sorting, predicates for whether a heading or table row should be displayed (to allow for only showing active keybindings), formatting of specific row entries, an update function for a row (e.g. to update a "previous definition" column), etc. A default "keybindings" type will be defined in the library, so all `evil-collection` would have to do is call the record and describe functions. The basic functionality is mostly implemented; I'll post a link when it's done.

> I just meant making general a dependency of evil-collection, and using the general key definers (with the added ability to specify who's making the the binding). I understand that general is user focused and much of this wouldn't be used by evil-collection, but I think it might be helpful to point people towards general as a useful package to use with evil.

I think either would potentially be fine, but even if `general-define-key` or a wrapper was used directly by evil-collection, some changes would need to be made (e.g. to prevent recording the keybindings as user keybindings). I'd prefer to have some separation even if within the same package (e.g. I could potentially add whitelisting/blacklisting for a package-focused definer). I can potentially work on making general more suitable for use in other packages in the future. I was against this in the past because I didn't think the syntactic convenience was worth using a non-standard definer from an external package, but for packages like evil-collection that are focused on keybindings and actually need extra functionality, I think it makes sense and could prevent code duplication.
